### PR TITLE
planner_cspace: improve peformance of DistanceMap

### PR DESCRIPTION
**Pull request by nhatao**
Screenshot from 2023-07-31 16-25-11
This image shows the performance analysis of `DistanceMap::fillCostmap()` using google perftools. Apparently, `unordered_map::operator[]` consumes too much computational time. 
Replacing `unordered_map` with `std::vector<bool>` brings significant performance improvement.

Example outputs in my local environment are the following.
Map size:  2799 x 995 x 16
Resolution: 0.05[m]
Distance between the start and the goal: about 120[m]

Cost estimation cache generated: about 32 times faster 
Cost estimation cache updated: about 25 times faster

**Debug outputs with the master branch**
```
[1690790535.233920154, 1690279770.184052384][/planner_3d]: New goal received (266, 513, 8)
[1690790537.351829088, 1690279772.293862314][/planner_3d]: Cost estimation cache generated (2.1179 sec.)
[1690790537.351967920, 1690279772.293862314][/planner_3d]: Navigation is stopping since the costmap is too old (costmap: 1690279770.064)
[1690790537.352054625, 1690279772.293862314][/planner_3d]: Map updated
[1690790537.353769621, 1690279772.293862314][/planner_3d]: Costmaps updated (0.0017)
[1690790537.859969914, 1690279772.803035835][/planner_3d]: Cost estimation cache search queue initial size: 46706, capacity: 60704
[1690790537.860035919, 1690279772.803035835][/planner_3d]: Cost estimation cache updated (0.5061 sec.)
[1690790537.860112541, 1690279772.803035835][/planner_3d]: Trying to publish message of type [geometry_msgs/PoseStamped/d3812c3cbc69362b77dc0b19b345f8f5] on a publisher with type [geometry_msgs/PoseStamped/d3812c3cbc69362b77dc0b19b345f8f5]
[1690790537.891392566, 1690279772.823144292][/planner_3d]: Path found (0.0312 sec.)      
[1690790537.891469260, 1690279772.823144292][/planner_3d]: Trying to publish message of type [geometry_msgs/PoseArray/916c28c5764443f268b296bb671b9d97] on a publisher with type [geometry_msgs/PoseArray/916c28c5764443f268b296bb671b9d97]
[1690790537.899781988, 1690279772.843365075][/planner_3d]: Hysteresis map generated (0.0081 sec.)
[1690790537.900435565, 1690279772.843365075][/planner_3d]: Map updated
[1690790537.902150747, 1690279772.843365075][/planner_3d]: Costmaps updated (0.0017)     
[1690790538.393336240, 1690279773.334053692][/planner_3d]: Cost estimation cache search queue initial size: 46696, capacity: 60704
[1690790538.393426504, 1690279773.334053692][/planner_3d]: Cost estimation cache updated (0.4911 sec.)
[1690790538.419920759, 1690279773.344108726][/planner_3d]: Path found (0.0263 sec.)      
[1690790538.427774427, 1690279773.371610907][/planner_3d]: Hysteresis map generated (0.0078 sec.)
[1690790538.428039656, 1690279773.371610907][/planner_3d]: Map updated
[1690790538.429667642, 1690279773.371610907][/planner_3d]: Costmaps updated (0.0016)     
[1690790538.923427120, 1690279773.866111335][/planner_3d]: Cost estimation cache search queue initial size: 46696, capacity: 60704
[1690790538.923461354, 1690279773.866111335][/planner_3d]: Cost estimation cache updated (0.4937 sec.)
[1690790538.949357555, 1690279773.876167141][/planner_3d]: Path found (0.0258 sec.)      
[1690790538.957408907, 1690279773.907101055][/planner_3d]: Hysteresis map generated (0.0078 sec.)
[1690790538.957689938, 1690279773.907101055][/planner_3d]: Map updated
[1690790538.959456506, 1690279773.907101055][/planner_3d]: Costmaps updated (0.0017)     
```

**Debug outputs with this PR**
```
[1690790751.789830462, 1690279754.931005985][/planner_3d]: New goal received (266, 513, 8)
[1690790751.855173026, 1690279754.989134325][/planner_3d]: Cost estimation cache generated (0.0653 sec.)
[1690790752.059618760, 1690279755.200362465][/planner_3d]: Map updated
[1690790752.061969536, 1690279755.200362465][/planner_3d]: Costmaps updated (0.0023)     
[1690790752.083231810, 1690279755.220471283][/planner_3d]: Cost estimation cache search queue initial size: 46703, capacity: 60704
[1690790752.083276022, 1690279755.220471283][/planner_3d]: Cost estimation cache updated (0.0212 sec.)
[1690790752.109915149, 1690279755.250654167][/planner_3d]: Trying to publish message of type [geometry_msgs/PoseStamped/d3812c3cbc69362b77dc0b19b345f8f5] on a publisher with type [geometry_msgs/PoseStamped/d3812c3cbc69362b77dc0b19b345f8f5]
[1690790752.139327576, 1690279755.270767577][/planner_3d]: Path found (0.0293 sec.)      
[1690790752.139409771, 1690279755.270767577][/planner_3d]: Trying to publish message of type [geometry_msgs/PoseArray/916c28c5764443f268b296bb671b9d97] on a publisher with type [geometry_msgs/PoseArray/916c28c5764443f268b296bb671b9d97]
[1690790752.147161750, 1690279755.280825338][/planner_3d]: Hysteresis map generated (0.0077 sec.)
[1690790752.208969795, 1690279755.349007466][/planner_3d]: Map updated  
[1690790752.211234635, 1690279755.349007466][/planner_3d]: Costmaps updated (0.0022)     
[1690790752.231968631, 1690279755.369077257][/planner_3d]: Cost estimation cache search queue initial size: 46662, capacity: 60704
[1690790752.231995987, 1690279755.369077257][/planner_3d]: Cost estimation cache updated (0.0207 sec.)
[1690790752.373863664, 1690279755.514019302][/planner_3d]: Map updated
[1690790752.375484502, 1690279755.514019302][/planner_3d]: Costmaps updated (0.0016)     
[1690790752.396173918, 1690279755.534159442][/planner_3d]: Cost estimation cache search queue initial size: 46628, capacity: 60704
[1690790752.396196714, 1690279755.534159442][/planner_3d]: Cost estimation cache updated (0.0207 sec.)
[1690790752.596777249, 1690279755.737620583][/planner_3d]: Map updated               
[1690790752.598622818, 1690279755.737620583][/planner_3d]: Costmaps updated (0.0018)     
[1690790752.619847926, 1690279755.757748560][/planner_3d]: Cost estimation cache search queue initial size: 46592, capacity: 60704
[1690790752.619873817, 1690279755.757748560][/planner_3d]: Cost estimation cache updated (0.0212 sec.)
[1690790752.645427245, 1690279755.777892843][/planner_3d]: Path found (0.0255 sec.)  
```

### CODE REVIEW

**Comment by at-wat** [comment]
## [466] PASSED on noetic

<details><summary>All tests passed</summary>


```
build/test_results/costmap_cspace/gtest-test_costmap_3d.xml: 22 tests
build/test_results/costmap_cspace/gtest-test_pointcloud_accumulator.xml: 4 tests
build/test_results/costmap_cspace/roslint-costmap_cspace.xml: 1 tests
build/test_results/joystick_interrupt/roslint-joystick_interrupt.xml: 1 tests
build/test_results/joystick_interrupt/rostest-test_test_joystick_interrupt_rostest.xml: 1 tests
build/test_results/joystick_interrupt/rosunit-test_joystick_interrupt.xml: 12 tests
build/test_results/map_organizer/roslint-map_organizer.xml: 1 tests
build/test_results/map_organizer/rostest-test_test_map_organizer_rostest.xml: 1 tests
build/test_results/map_organizer/rostest-test_test_pointcloud_to_maps_rostest.xml: 1 tests
build/test_results/map_organizer/rosunit-test_map_organizer.xml: 8 tests
build/test_results/map_organizer/rosunit-test_pointcloud_to_maps.xml: 2 tests
build/test_results/neonavigation_common/roslint-neonavigation_common.xml: 1 tests
build/test_results/neonavigation_common/rostest-test_test_compat_rostest.xml: 1 tests
build/test_results/neonavigation_common/rosunit-test_compat.xml: 6 tests
build/test_results/neonavigation_metrics_msgs/gtest-test_helper.xml: 4 tests
build/test_results/neonavigation_metrics_msgs/roslint-neonavigation_metrics_msgs.xml: 1 tests
build/test_results/obj_to_pointcloud/roslint-obj_to_pointcloud.xml: 1 tests
build/test_results/obj_to_pointcloud/rostest-test_test_obj_to_pointcloud_rostest.xml: 1 tests
build/test_results/obj_to_pointcloud/rosunit-test_obj_to_pointcloud.xml: 2 tests
build/test_results/planner_cspace/gtest-test_blockmem_gridmap.xml: 10 tests
build/test_results/planner_cspace/gtest-test_costmap_bbf.xml: 4 tests
build/test_results/planner_cspace/gtest-test_cyclic_vec.xml: 14 tests
build/test_results/planner_cspace/gtest-test_distance_map.xml: 14 tests
build/test_results/planner_cspace/gtest-test_distance_map_fast_update.xml: 4 tests
build/test_results/planner_cspace/gtest-test_grid_astar.xml: 12 tests
build/test_results/planner_cspace/gtest-test_grid_metric_converter.xml: 4 tests
build/test_results/planner_cspace/gtest-test_motion_cache.xml: 2 tests
build/test_results/planner_cspace/gtest-test_motion_primitive_builder.xml: 2 tests
build/test_results/planner_cspace/gtest-test_path_interpolator.xml: 4 tests
build/test_results/planner_cspace/gtest-test_planner_3d_cost.xml: 2 tests
build/test_results/planner_cspace/roslint-planner_cspace.xml: 1 tests
build/test_results/planner_cspace/rostest-navigation_rostest__antialias_start_true.xml: 1 tests
build/test_results/planner_cspace/rostest-navigation_rostest__antialias_start_true__fast_map_update_true.xml: 1 tests
build/test_results/planner_cspace/rostest-navigation_rostest__with_tolerance_true.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_abort_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_costmap_watchdog_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_debug_outputs_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_navigation_boundary_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_navigation_compat_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_navigation_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_planner_2dof_serial_joints_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_planner_3d_map_size_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_preempt_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_tolerant_action_rostest.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_abort.xml: 2 tests
build/test_results/planner_cspace/rosunit-test_costmap_watchdog.xml: 4 tests
build/test_results/planner_cspace/rosunit-test_debug_outputs.xml: 8 tests
build/test_results/planner_cspace/rosunit-test_navigate.xml: 12 tests
build/test_results/planner_cspace/rosunit-test_navigate_boundary.xml: 2 tests
build/test_results/planner_cspace/rosunit-test_planner_2dof_serial_joints.xml: 4 tests
build/test_results/planner_cspace/rosunit-test_planner_3d_map_size.xml: 12 tests
build/test_results/planner_cspace/rosunit-test_preempt.xml: 2 tests
build/test_results/planner_cspace/rosunit-test_tolerant_action.xml: 2 tests
build/test_results/safety_limiter/roslint-safety_limiter.xml: 1 tests
build/test_results/safety_limiter/rostest-test_test_safety_limiter2_rostest.xml: 1 tests
build/test_results/safety_limiter/rostest-test_test_safety_limiter_compat_rostest.xml: 1 tests
build/test_results/safety_limiter/rostest-test_test_safety_limiter_rostest.xml: 1 tests
build/test_results/safety_limiter/rosunit-test_safety_limiter.xml: 22 tests
build/test_results/safety_limiter/rosunit-test_safety_limiter2.xml: 2 tests
build/test_results/track_odometry/gtest-test_tf_projection.xml: 2 tests
build/test_results/track_odometry/roslint-track_odometry.xml: 1 tests
build/test_results/track_odometry/rostest-test_test_tf_projection_rostest.xml: 1 tests
build/test_results/track_odometry/rostest-test_test_track_odometry_rostest.xml: 1 tests
build/test_results/track_odometry/rosunit-test_tf_projection_node.xml: 8 tests
build/test_results/track_odometry/rosunit-test_track_odometry.xml: 10 tests
build/test_results/trajectory_tracker/gtest-test_trajectory_tracker_filter.xml: 6 tests
build/test_results/trajectory_tracker/gtest-test_trajectory_tracker_path2d.xml: 8 tests
build/test_results/trajectory_tracker/roslint-trajectory_tracker.xml: 1 tests
build/test_results/trajectory_tracker/rostest-test_test_trajectory_recorder_rostest.xml: 1 tests
build/test_results/trajectory_tracker/rostest-test_test_trajectory_tracker_overshoot_rostest.xml: 1 tests
build/test_results/trajectory_tracker/rostest-test_test_trajectory_tracker_rostest.xml: 1 tests
build/test_results/trajectory_tracker/rostest-test_test_trajectory_tracker_with_odom_rostest.xml: 1 tests
build/test_results/trajectory_tracker/rostest-trajectory_tracker_rostest__odom_delay_0.xml: 1 tests
build/test_results/trajectory_tracker/rosunit-test_trajectory_recorder.xml: 2 tests
build/test_results/trajectory_tracker/rosunit-test_trajectory_tracker.xml: 16 tests
build/test_results/trajectory_tracker/rosunit-test_trajectory_tracker_overshoot.xml: 12 tests
build/test_results/trajectory_tracker/rosunit-test_trajectory_tracker_with_odom.xml: 4 tests
build/test_results/trajectory_tracker_msgs/gtest-test_path_with_velocity_conversion.xml: 4 tests
build/test_results/trajectory_tracker_msgs/roslint-trajectory_tracker_msgs.xml: 1 tests
Summary: 315 tests, 0 errors, 0 failures, 0 skipped
```
</details>

File: planner_cspace/src/distance_map.cpp [hunk 1]
```diff
@@ -138,7 +136,7 @@ void DistanceMap::fillCostmap(
             }
             if (collision)
             {
-              edges_local[p] = true;
+              edges_buf_[p[0] + p[1] * p_.size[0]] = true;
```
Comment (at-wat) [inline planner_cspace/src/distance_map.cpp:139]: https://en.cppreference.com/w/cpp/container/vector_bool
> Does not guarantee that different elements in the same container can be modified concurrently by different threads. 

This line is executed by multiple threads

Comment (nhatao) [reply planner_cspace/src/distance_map.cpp]: I added temporal storage named `std::vector<size_t> edge_indexes`.

Comment (nhatao) [reply planner_cspace/src/distance_map.cpp]: I think the performance degradation is acceptable.
```
[1690797074.360340821, 1690279749.841883353][/planner_3d]: New goal received (266, 513, 8)
1690797074.424991012, 1690279749.902270411][/planner_3d]: Cost estimation cache generated (0.0646 sec.)
[1690797074.431511267, 1690279749.912324169][/planner_3d]: Map updated
[1690797074.433128297, 1690279749.912324169][/planner_3d]: Costmaps updated (0.0016)
[1690797074.455404871, 1690279749.932516006][/planner_3d]: Cost estimation cache search queue initial size: 46695, capacity: 60704
[1690797074.455429775, 1690279749.932516006][/planner_3d]: Cost estimation cache updated (0.0222 sec.)
[1690797074.471341985, 1690279749.952620123][/planner_3d]: Trying to publish message of type [geometry_msgs/PoseStamped/d3812c3cbc69362b77dc0b19b345f8f5] on a publisher with type [geometry_msgs/PoseStamped/d3812c3cbc69362b77dc0b19b345f8f5]
[1690797074.501594817, 1690279749.972727846][/planner_3d]: Path found (0.0302 sec.)     
[1690797074.501627082, 1690279749.972727846][/planner_3d]: Trying to publish message of type [geometry_msgs/PoseArray/916c28c5764443f268b296bb671b9d97] on a publisher with type [geometry_msgs/PoseArray/916c28c5764443f268b296bb671b9d97]
[1690797074.508748404, 1690279749.972727846][/planner_3d]: Hysteresis map generated (0.0071 sec.)
[1690797074.556863731, 1690279750.037711948][/planner_3d]: Map updated
[1690797074.559274794, 1690279750.037711948][/planner_3d]: Costmaps updated (0.0023)
[1690797074.580336898, 1690279750.057821310][/planner_3d]: Cost estimation cache search queue initial size: 46685, capacity: 60704
[1690797074.580366817, 1690279750.057821310][/planner_3d]: Cost estimation cache updated (0.0210 sec.)
[1690797074.756607771, 1690279750.238098679][/planner_3d]: Map updated
[1690797074.758966322, 1690279750.238098679][/planner_3d]: Costmaps updated (0.0023)    
[1690797074.781845387, 1690279750.258207917][/planner_3d]: Cost estimation cache search queue initial size: 46685, capacity: 60704
[1690797074.781878368, 1690279750.258207917][/planner_3d]: Cost estimation cache updated (0.0228 sec.)
[1690797074.999613759, 1690279750.475198153][/planner_3d]: Path found (0.0260 sec.)     
[1690797075.008093280, 1690279750.485255865][/planner_3d]: Hysteresis map generated (0.0084 sec.)
[1690797075.008279763, 1690279750.485255865][/planner_3d]: Map updated
[1690797075.009892748, 1690279750.485255865][/planner_3d]: Costmaps updated (0.0016)    
[1690797075.031660764, 1690279750.505362525][/planner_3d]: Cost estimation cache search queue initial size: 46685, capacity: 60704
[1690797075.031693066, 1690279750.505362525][/planner_3d]: Cost estimation cache updated (0.0217 sec.)

```

**Review by at-wat — APPROVED** [review]
LGTM

**Pull request merged**